File: generate_scatter.py
import matplotlib.pyplot as plt
import numpy as np
import os

# Ensure img dir exists
os.makedirs('img', exist_ok=True)

# Use clean default style
plt.style.use('default')
fig, ax = plt.subplots(figsize=(10, 6.2), facecolor='white')
ax.set_facecolor('#f8f9fa')

# Data: [Range (m), Bandwidth (bps), Power (mA scale)]
techs = {
    'Wi-Fi': [50, 1e8, 500],
    'Bluetooth': [10, 2e6, 80],
    'BLE': [30, 1e6, 15],
    'Zigbee/Z-Wave': [100, 250000, 20],
    'LoRaWAN': [15000, 5000, 25],
    'NB-IoT': [20000, 200000, 150],
    'Sigfox': [40000, 100, 15],
    '5G (Cellular)': [5000, 1e9, 800]
}

x = [v[0] for v in techs.values()]
y = [v[1] for v in techs.values()]
sizes = [v[2]*1.5 for v in techs.values()] # scale for bubble sizes

# Create scatter plot
scatter = ax.scatter(x, y, s=sizes, alpha=0.85, c=[v[2] for v in techs.values()], cmap='YlOrRd', edgecolors='#111111', linewidths=1.5)

# Annotate each technology label in bold black
for i, txt in enumerate(techs.keys()):
    ax.annotate(txt, (x[i], y[i]), xytext=(12, 0), textcoords='offset points', color='black', fontsize=11, fontweight='bold', va='center')

# Log scales as required by the logarithmic nature of range/bandwidth
ax.set_xscale('log')
ax.set_yscale('log')

# Axis labels and titles in bold black (bilingual)
ax.set_xlabel('Zasięg / Range [m]', fontsize=13, fontweight='bold', color='black', labelpad=10)
ax.set_ylabel('Przepustowość / Bandwidth [b/s]', fontsize=13, fontweight='bold', color='black', labelpad=10)
ax.set_title('IoT Connectivity: Range vs. Bandwidth vs. Power', fontsize=16, fontweight='bold', pad=20, color='black')

# Style ticks
ax.tick_params(colors='black', which='both', labelsize=10)

# Add a colorbar to explain size/color with bold black labels
cbar = plt.colorbar(scatter)
cbar.set_label('Power Consumption / Pobór prądu [mA]', fontsize=11, fontweight='bold', color='black', labelpad=10)
cbar.ax.tick_params(colors='black')

# Grid setup
plt.grid(True, which="both", ls="--", alpha=0.5, color='#cccccc')
plt.tight_layout()

# Save with a solid white background (not transparent) so it is universally high contrast
plt.savefig('img/scatter_triangle.png', transparent=False, dpi=300, facecolor='white')
print("Saved light-themed scatter plot to img/scatter_triangle.png")

File: generate_scatter.R
library(ggplot2)

# Ensure img directory exists
dir.create("img", showWarnings = FALSE)

# Data definition
techs <- data.frame(
  Name = c('Wi-Fi', 'Bluetooth', 'BLE', 'Zigbee/Z-Wave', 'LoRaWAN', 'NB-IoT', 'Sigfox', '5G (Cellular)'),
  Range = c(50, 10, 30, 100, 15000, 20000, 40000, 5000),
  Bandwidth = c(1e8, 2e6, 1e6, 250000, 5000, 200000, 100, 1e9),
  Power = c(500, 80, 15, 20, 25, 150, 15, 800)
)

# Plotting
p <- ggplot(techs, aes(x = Range, y = Bandwidth)) +
  # Grid lines (major and minor)
  theme_bw() +
  # Bubbles with outline and color scale based on Power consumption
  geom_point(aes(fill = Power, size = Power), shape = 21, color = "#111111", stroke = 1.5, alpha = 0.85) +
  # Custom gradient for hot-to-cold power consumption
  scale_fill_gradientn(
    colors = c("#2ca02c", "#ff7f0e", "#d62728"), # green to orange to red
    name = "Power Consumption /\nPobór prądu [mA]"
  ) +
  # Adjust bubble sizes nicely
  scale_size_continuous(range = c(6, 20), guide = "none") + 
  # Log scale for Range (X axis) with beautiful labels
  scale_x_log10(
    breaks = c(10, 100, 1000, 10000, 100000),
    labels = c("10 m", "100 m", "1 km", "10 km", "100 km")
  ) +
  # Log scale for Bandwidth (Y axis) with beautiful labels
  scale_y_log10(
    breaks = c(100, 1000, 10000, 100000, 1e6, 1e7, 1e8, 1e9),
    labels = c("100 b/s", "1 kb/s", "10 kb/s", "100 kb/s", "1 Mb/s", "10 Mb/s", "100 Mb/s", "1 Gb/s")
  ) +
  # Bold, black labels for each protocol
  geom_text(
    aes(label = Name), 
    fontface = "bold", 
    color = "black", 
    size = 4.2, 
    vjust = -1.5, 
    hjust = 0.5
  ) +
  # Axis labels and Title
  labs(
    title = "IoT Connectivity: Range vs. Bandwidth vs. Power",
    subtitle = "Złoty Trójkąt Łączności IoT: Kompromisy",
    x = "Zasięg / Range [Log]",
    y = "Przepustowość / Bandwidth (Throughput) [Log]"
  ) +
  # Clean high-contrast theme styling
  theme(
    plot.title = element_text(face = "bold", size = 16, color = "black", hjust = 0.5, margin = margin(b = 5)),
    plot.subtitle = element_text(size = 12, color = "#444444", hjust = 0.5, face = "italic", margin = margin(b = 15)),
    axis.title.x = element_text(face = "bold", size = 12, color = "black", margin = margin(t = 10)),
    axis.title.y = element_text(face = "bold", size = 12, color = "black", margin = margin(r = 10)),
    axis.text = element_text(color = "black", size = 10, face = "bold"),
    legend.title = element_text(face = "bold", size = 10, color = "black"),
    legend.text = element_text(color = "black", size = 9, face = "bold"),
    panel.grid.major = element_line(color = "#cccccc", linetype = "dashed"),
    panel.grid.minor = element_line(color = "#e5e5e5", linetype = "dotted"),
    panel.background = element_rect(fill = "#f8f9fa"),
    plot.background = element_rect(fill = "white", color = NA)
  )

# Save plot as high-resolution PNG with solid white background
ggsave("img/scatter_triangle.png", plot = p, width = 10, height = 6.2, dpi = 300, bg = "white")
cat("Successfully generated high-contrast, bilingual scatter plot in img/scatter_triangle.png using R ggplot2!\n")
